### PR TITLE
Jakarta Pages 4.0.0

### PR DESCRIPTION
**Pull request by markt-asf**
## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [x] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [x] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [x] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [X] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - ~[ ] Generated IP Log~ *no longer required according to project handbook*
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [X] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/repositories/jakartaservletjsp-1031
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/pages/jakartaee11/staged/eftl/jakarta-pages-tck-4.0.0.zip
- [X] The URL of the compatibility certification request issue:
      https://github.com/jakartaee/pages/issues/271
- [x] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by markt-asf** [comment]
This is an initial draft. Everything is complete and ready for the release review - I just need to ensure everything is staged to the right place and complete the paperwork.

**Comment by alwin-joseph** [comment]
@markt-asf The TCK https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee11/staged/eftl/jakarta-pages-tck-4.0.0.zip was the one created from platform tck project before the migration of the Server Pages TCK.

I would like to remove this old version once a new Jenkins job from https://ci.eclipse.org/jsp  builds and copies a TCK bundle at https://download.eclipse.org/ee4j/pages/ using the instructions given at  https://github.com/jakartaee/pages/pull/261#issuecomment-2026751030. Please let me know if you are facing any issue with this.

**Comment by markt-asf** [comment]
Note: The TCK being used still has SNAPSHOT dependencies. This review should not progress until an updated TCK is available without SNAPSHOT dependencies.

**Comment by ivargrimstad** [comment]
# Mentor Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [x] N/A (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation
  - [ ] Review summary statement to ensure it has been updated to reflect the work accomplished and does not contain proposal statements as might be found in Plan Review statement.

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request

10. If a Release Review is required, the specification project team contacts the EMO to initiate the review by sending an email to emo@eclipse.org.
    (A Release Review is not required if the current release is a Service Release based on a previously successful Major or Minor
    release as indicated by a release record on the project's Releases page, e.g., the [Jakarta Servlet releases page](https://projects.eclipse.org/projects/ee4j.servlet/reviews).)
    - [ ] An issue will be created by the EMO to track the release review.
    - [ ] The specification project team requests approval for the release from the PMC by sending an email to ee4j-pmc@eclipse.org.
    - [ ] The specification project team then delivers an IP Log to the IP Team for their review as described in https://www.eclipse.org/projects/handbook/#pmi-commands-iplog.

11. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by mtdelgadoa** [comment]
## EMO REVIEW CHECKLIST 

EDP Review status: ONGOING

### **EMO review checklist**
PMI record: [URL](https://projects.eclipse.org/projects/ee4j.jsp/reviews/4.0.0-release-review)

_**EF Specification Process**_      
- [ ] Spec Committee Ballot completed

_**Intellectual Property Management**_      
- [ ] All project code has [copyright and license headers](https://www.eclipse.org/projects/handbook/#ip-copyright-headers) correctly applied. ** EMO will scan the code at their discretion **   
- [ ]  All distributed third-party content has been vetted by the IP Due Diligence process (i.e., IP Log has been approved)   

_**Open Source Rules of Engagement**_
- [ ] PMC approval [on the mailing list](https://github.com/eclipsefdn/emo-team/emo/-/blob/main/%7Blink%20to%20the%20message%7D)

General:
- [ ] Project is operating within the mission and scope defined in its top-level project’s charter   
- [ ]  Project is operating within the bounds of its own scope   
- [ ]  Project is operating in an open and transparent manner       
- [ ]  Overall the project is operating according to the Eclipse Development Process.   

Things to check:
- Communication channels advertised
- Advertised communication channels used
- Committers are responding to questions
- Committers are responding to issues
- Committers are responding to pull/merge/review requests

_**Branding and Trademarks**_
The following applies when the project has a custom website.
To the best of our knowledge:
- [ ]  Project content correctly uses Eclipse Foundation trademarks    
- [ ] Project content (code and documentation) does not violate trademarks owned by other organizations   

Things to check:
- [ ]  Project website uses the [project's formal name](https://www.eclipse.org/projects/handbook/#trademarks-website-name) in first and all prominent references
- [ ]   Project website includes a trademark attribution statement
- [ ] Project website footers contain all [necessary elements](https://www.eclipse.org/projects/handbook/#trademarks-website-footer)

_**Legal Documentation**_
Required files:   
- [ ] License files in all repository roots   
- [ ]  README       
- [ ]  CONTRIBUTING (or equivalent)   

Recommended files:
- [ ] NOTICES or equivalent (you can use the [Legal Documentation generator](https://www.eclipse.org/projects/tools/documentation.php) available under committer tools)   
- [ ]  CODE_OF_CONDUCT   
- [ ]  SECURITY   
 
See examples for [Security file](https://gitlab.eclipse.org/security/best-practices/-/blob/main/templates/SECURITY.md) and [Code of Conduct](https://github.com/eclipse/.github/blob/master/CODE_OF_CONDUCT.md).

Required elements:
- [ ]  ECA is referenced/described   

Recommended elements:
- [ ]  Build instructions provided   
- [ ] [Security policy](https://www.eclipse.org/security/policy.php) is described   
 
_**Metadata (PMI)**_
- [ ] The formal name, e.g. "Eclipse Foo™", is used in the project title   
- [ ] The formal name including appropriate marks (e.g, "™") is used in the first mention in the text of the project description, and scope   
- [ ] The project description starts with a single paragraph that can serve as an executive summary   
- [ ] Source code repository references are up-to-date   
- [ ] Download links and information are up-to-date (see [EF handbook for more information on how-to do this](https://www.eclipse.org/projects/handbook/#resources-downloads))   
- [ ] Communication channels listed in the PMI (i.e. public mailing list, forums, etc.)

**Comment by alwin-joseph** [comment]
> Note: The TCK being used still has SNAPSHOT dependencies. This review should not progress until an updated TCK is available without SNAPSHOT dependencies.

I have updated the TCK at https://www.eclipse.org/downloads/download.php?file=/ee4j/jakartaee-tck/jakartaee11/staged/eftl/jakarta-pages-tck-4.0.0.zip to remove the SNAPSHOT dependencies for now. This was built using the Jenkins job https://ci.eclipse.org/jakartaee-tck/job/11/job/standalone-tck/job/eftl-pages-standalonetck-build_from_pages_project/ that builds TCK from [jakartaee/pages/tck](https://github.com/jakartaee/pages/tree/master/tck) .

**Comment by keilw** [comment]
@markt-asf @ivargrimstad @arjantijms I could not find a Compatibility certification link under https://github.com/eclipse-ee4j/wasp/issues, since Tomcat 11 is mentioned as the compatible implementation, could that be outside Eclipse?

**Comment by keilw** [comment]
@markt-asf @ivargrimstad There is the name "Apache Tomcat 11.0.0-M19-SNAPSHOT" of a compatible implementation, but no link, is that good enough?

**Comment by keilw** [comment]
@markt-asf About the javadocs, while there may not be many changes aside from the Removal of deprecated API, was there a reason to stick with "Copyright © 2018, 2021 Eclipse Foundation." in the footer instead of ",2023" or ",2024"?

**Comment by arjantijms** [comment]
> @markt-asf @ivargrimstad @arjantijms I could not find a Compatibility certification link under https://github.com/eclipse-ee4j/wasp/issues, since Tomcat 11 is mentioned as the compatible implementation, could that be outside Eclipse?

It's perfectly fine that an implementation outside eclipse ratifies as the compatible implementation.

WaSP will certify at a later date.

**Comment by ivargrimstad** [comment]
> > @markt-asf @ivargrimstad @arjantijms I could not find a Compatibility certification link under https://github.com/eclipse-ee4j/wasp/issues, since Tomcat 11 is mentioned as the compatible implementation, could that be outside Eclipse?
> 
> It's perfectly fine that an implementation outside eclipse ratifies as the compatible implementation.
> 
> WaSP will certify at a later date.

Yes, the more implementations the merrier!

**Comment by ivargrimstad** [comment]
> @markt-asf @ivargrimstad There is the name "Apache Tomcat 11.0.0-M19-SNAPSHOT" of a compatible implementation in the `_index.md`, but no link, is that good enough?

There should be a link to a location where it is possible to download the implementation

**Review by keilw — CHANGES_REQUESTED** [review]
@markt-asf Could you please 

1. Add the download link for Apache Tomcat 11.0.0-M19-SNAPSHOT to /pages/4.0/_index.md
2. Point to the Compatibility certification link of Apache Tomcat 11.0.0
3. And ideally change the footer in the apidocs to "Copyright © 2018, 2024 Eclipse Foundation."

**Comment by markt-asf** [comment]
Progressing the PR will require an updated Pages TCK to be staged. The current one has a missing dependency. It is also using unexpected Maven coordinates.

If I hack my way around the issues with the Pages TCK then the tests all pass so I should be able to progress this PR quickly once an updated TCK is available.

Ah. The Pages TCK has been transferred to the Pages project. That essentially means I am going to have to figure out how to fix this (it looks as if it is already fixed) and how to stage an updated TCK (may not be quite so easy). Not sure how long any of this is going to take.

Since it is a relatively small change to hack around the TCK issues (add a dependency, change Maven coordinates) we could proceed with the 2024-04-16 version of the TCK. The main thing I don't like about that is that the finally released TCK will be wrong.

I am going to see if I can get a fixed version of the TCK staged. If that looks to be taking too long, the fallback will be to use the 2024-04-16 version and hack around the issues.

**Comment by markt-asf** [comment]
With some much appreciated help from @arjantijms, I have been able to stage the updated Pages TCK via Jenkins. There are still some niggles to be ironed out but I am confident I'll be able to address those quickly. I expect to be able to stage the final Pages JCK, re-stage the Pages API with the copyright fix for the API docs and provided an updated CCR using Tomcat 11.0.0-M20 by the end of today.

**Comment by alwin-joseph** [comment]
> With some much appreciated help from @arjantijms, I have been able to stage the updated Pages TCK via Jenkins. There are still some niggles to be ironed out but I am confident I'll be able to address those quickly. I expect to be able to stage the final Pages JCK, re-stage the Pages API with the copyright fix for the API docs and provided an updated CCR using Tomcat 11.0.0-M20 by the end of today.

I see a staged version of pages tck at https://download.eclipse.org/ee4j/pages/jakartaee11/staged/eftl/ 
Can you please add below also to the jenkins job to build the userguide pdf/html. 
`cd $WORKSPACE/tck/docs; mvn`
`cd $WORKSPACE/tck; mvn clean install`

**Comment by markt-asf** [comment]
@alwin-joseph Thanks. User guide should be fixed now and I expect the mirrors will pick up the new version fairly quickly.

**Comment by markt-asf** [comment]
I think that is all the requested changes completed. Pages 4.0.0 is (hopefully) ready for the release ballot now.

File: pages/4.0/_index.md [hunk 1]
```diff
@@ -19,27 +19,21 @@ This release removes deprecated code and provides any additional updates necessa
...
-* [Jakarta Pages 3.1 Specification Document](./jakarta-server-pages-spec-3.1.html) (HTML)
-* [Jakarta Pages 3.1 Javadoc](./apidocs)
-* [Jakarta Pages 3.1 TCK](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip)([sig](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip.sha256),[pub](https://jakarta.ee/specifications/jakartaee-spec-committee.pub))
+  * [Jakarta EE Platform 11 Release Plan](https://jakartaee.github.io/platform/jakartaee11/JakartaEE11ReleasePlan)
+* [Jakarta Pages 4.0 Specification Document](./jakarta-server-pages-spec-4.0.pdf) (PDF)
+* [Jakarta Pages 4.0 Specification Document](./jakarta-server-pages-spec-4.0.html) (HTML)
+* [Jakarta Pages 4.0 Javadoc](./apidocs)
+* [Jakarta Pages 4.0 TCK](https://download.eclipse.org/ee4j/pages/jakartaee11/staged/eftl/jakarta-pages-tck-4.0.0.zip), sig (not avaialble),[sha](https://download.eclipse.org/ee4j/pages/jakartaee11/staged/eftl/jakarta-pages-tck-4.0.0.zip.info),[pub](https://jakarta.ee/specifications/jakartaee-spec-committee.pub))
```
Comment (Pandrex247) [inline pages/4.0/_index.md:30]: This should be set to the final download location, not the staged location.
```suggestion
* [Jakarta Pages 4.0 TCK](https://download.eclipse.org/jakartaee/pages/4.0/jakarta-pages-tck-4.0.0.zip), ([sig](https://download.eclipse.org/jakartaee/pages/4.0/jakarta-pages-tck-4.0.0.zip.sig), [sha](https://download.eclipse.org/jakartaee/pages/4.0/jakarta-pages-tck-4.0.0.zip.sha256), [pub](https://jakarta.ee/specifications/jakartaee-spec-committee.pub))
```

File: pages/4.0/_index.md [hunk 2]
```diff
@@ -19,27 +19,21 @@ This release removes deprecated code and provides any additional updates necessa
...
-<!--
-
-The following can be uncommented and version information updated as they become available.
-
-  * [Jakarta EE Platform 10 Release Plan](https://jakartaee.github.io/platform/jakartaee10/JakartaEE10ReleasePlan)
-* [Jakarta Pages 3.1 Specification Document](./jakarta-server-pages-spec-3.1.pdf) (PDF)
-* [Jakarta Pages 3.1 Specification Document](./jakarta-server-pages-spec-3.1.html) (HTML)
-* [Jakarta Pages 3.1 Javadoc](./apidocs)
-* [Jakarta Pages 3.1 TCK](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip)([sig](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/pages/3.1/jakarta-pages-tck-3.1.0.zip.sha256),[pub](https://jakarta.ee/specifications/jakartaee-spec-committee.pub))
+  * [Jakarta EE Platform 11 Release Plan](https://jakartaee.github.io/platform/jakartaee11/JakartaEE11ReleasePlan)
```
Comment (Pandrex247) [inline pages/4.0/_index.md:26]: Missing link to the Pages 4.0 plan
```suggestion
* [Jakarta Server Pages 4.0 Release Plan](https://projects.eclipse.org/projects/ee4j.jsp/releases/4.0.0/plan)
  * [Jakarta EE Platform 11 Release Plan](https://jakartaee.github.io/platform/jakartaee11/JakartaEE11ReleasePlan)
```

**Review by keilw — APPROVED** [review]

**Comment by ivargrimstad** [comment]
On ballot completion, the specification committee mentor (@keilw):

 - [x] adds this final checklist to the main PR.
 - [ ] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [ ] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
https://github.com/jakartaee/servlet/issues/651

**Pull request merged**